## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [94]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [95]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [96]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [97]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#df.index
#df.shape
#df.dtypes


In [98]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [99]:
# For the category with the most entries,
# which subcategory had the most entries?
top_subcategory = df.loc[df['category'] == 'consumables']['subcategory'].value_counts().head(1)
top_subcategory

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [100]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [101]:
# Store the client ids of those top 5 clients in a list.
top_clients_ids = top_clients.index.tolist()
top_clients_ids



[33615, 66037, 46820, 38378, 24741]

In [102]:
# How many total units (the qty column) did the
# client with the most entries order order?
print(df.loc[df['client_id'] == 33615]['qty'].sum())


64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [103]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']

In [104]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['shipping_price'] = df.apply(lambda x: x['unit_weight'] * x['qty'] * 7 if x['unit_weight'] > 50 else x['unit_weight'] * x['qty'] * 10, axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0          7875.0  
1           312.9  
2           655.2  
3           356.7  
4          9286.0

In [105]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price'] + ((df['subtotal'] + df['shipping_price']) * 0.0925)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00          7875.0   
1      15.09         1.49   21            0     523.95           312.9   
2       7.86         1.68   39            6     527.28           655.2   
3      24.85         1.23   29            3    1056.18           356.7   
4     108.17        46.43   20            1    3902.00          9286.0   

     total_price  
0  134420.107500  
1     914.258625  
2    1291.859400  
3    1543.571400  
4   14407.890000  

[5 rows x 21 columns]

In [106]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00          7875.0  134420.107500   
1         1.49   21            0     523.95           312.9     914.258625   
2         1.68   39            6     527.28           655.2    1291.859400   
3         1.23   29            3    1056.18           356.7    1543.571400   
4        46.43   20            1    3902.00          9286.0   14407.890000   

   total_cost  
0    87959.55  
1      629.79  
2      961.74  
3     1077.35  
4    11449.40  

[5 rows x 22 columns]

In [107]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['total_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  total_cost  \
0            1  115164.00          7875.0  134420.107500    87959.55   
1            0     523.95           312.9     914.258625      629.79   
2            6     527.28           655.2    1291.859400      961.74   
3            3    1056.18           356.7    1543.571400     1077.35   
4            1    3902.00          9286.0   14407.890000    11449.40   

         profit  
0  46460.557500  
1    284.468625  
2    330.119400  
3    466.221400  
4   2958.490000  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [108]:
# Calculate the total price for each order and compare it with the provided total prices
order_totals = df.groupby('order_id')['total_price'].sum()
order_ids = [2742071, 2173913, 6128929]
order_totals.loc[order_ids]



order_id
2742071    158105.649525
2173913    180488.057550
6128929    996593.762225
Name: total_price, dtype: float64

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [109]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients_spending = df[df['client_id'].isin(top_clients_ids)].groupby('client_id')['total_price'].sum()
top_clients_spending


client_id
24741    8.466399e+07
33615    9.228995e+06
38378    1.450805e+07
46820    1.048972e+07
66037    1.090884e+07
Name: total_price, dtype: float64

In [110]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = pd.DataFrame({
    'total_units_purchased': df[df['client_id'].isin(top_clients_ids)].groupby('client_id')['qty'].sum(),
    'total_shipping_price': df[df['client_id'].isin(top_clients_ids)].groupby('client_id')['shipping_price'].sum(),
    'total_revenue': df[df['client_id'].isin(top_clients_ids)].groupby('client_id')['total_price'].sum(),
    'total_profit': df[df['client_id'].isin(top_clients_ids)].groupby('client_id')['profit'].sum()
})
summary_df.sort_values('total_profit', ascending=False)

total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862             7318757.5   8.466399e+07   
38378                      73667             4895361.6   1.450805e+07   
66037                      43018             1989501.5   1.090884e+07   
46820                      75768             2284215.0   1.048972e+07   
33615                      64313             2608560.3   9.228995e+06   

           total_profit  
client_id                
24741      3.678278e+07  
38378      3.407426e+06  
66037      3.310009e+06  
46820      2.799759e+06  
33615      2.274105e+06

In [119]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
# Divide values in total shipping price, total revenue, and total profit columns by 1,000,000 and name them accordingly.
summary_df['total_shipping_price'] = summary_df['total_shipping_price'] / 1000000
summary_df['total_revenue'] = summary_df['total_revenue'] / 1000000
summary_df['total_profit'] = summary_df['total_profit'] / 1000000
summary_df = summary_df.rename(columns={'total_units_purchased': 'Total Units Purchased', 'total_shipping_price': 'Total Shipping Price (Millions)', 'total_revenue': 'Total Revenue (Millions)', 'total_profit': 'Total Profit (Millions)'})
summary_df



KeyError: 'total_shipping_price'

In [120]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df.sort_values('Total Profit (Millions)', ascending=False)

Total Units Purchased  Total Shipping Price (Millions)  \
client_id                                                           
24741                     239862                         0.000007   
38378                      73667                         0.000005   
66037                      43018                         0.000002   
46820                      75768                         0.000002   
33615                      64313                         0.000003   

           Total Revenue (Millions)  Total Profit (Millions)  
client_id                                                     
24741                     84.663990                36.782781  
38378                     14.508053                 3.407426  
66037                     10.908842                 3.310009  
46820                     10.489716                 2.799759  
33615                      9.228995                 2.274105